### PR TITLE
Task-5: BE: Implement create user(sign up) Post API

### PR DESCRIPTION
**Pull request by denys-rybachok**

### CODE REVIEW

File: src/entities/user/service/user.service.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import {Injectable, HttpException, HttpStatus} from '@nestjs/common';
+import {UserDocument} from '../model/user.schema';
+import {InjectModel} from '@nestjs/mongoose';
+import {Model} from 'mongoose';
+import {statusMessages} from '@app/constants/errorMessages.constant';
+import * as bcrypt from 'bcrypt';
+import * as jwt from 'jsonwebtoken';
+import {CreateUserDto} from '../dto/createUserDto';
+
+export type UserWithoutPasswordHash = Omit<UserDocument, 'passwordHash'>;
```
Comment (VYaniv) [inline src/entities/user/service/user.service.ts]: Put in in separate file for types or models.

File: src/entities/user/model/user.schema.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  firstName: string;
+
+  @Prop({required: true})
+  lastName: string;
+
+  @Prop({required: true})
+  email: string;
+
+  @Prop({required: true})
+  passwordHash: string;
```
Comment (VYaniv) [inline src/entities/user/model/user.schema.ts]: `passwordHash` -> `password` . It doesn`t make a sense to rename it to hash, just transform the value but keep the same key.

File: src/entities/user/validation/createUserValidation.schema.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import * as yup from 'yup';
+import {validationErrors} from '@app/constants/validationErrors.constant';
+
+export const createUserSchema = yup.object().shape({
+  firstName: yup.string().required(validationErrors.FIELD_REQUIRED),
+  lastName: yup.string().required(validationErrors.FIELD_REQUIRED),
+  email: yup.string().required(validationErrors.FIELD_REQUIRED).email(validationErrors.INVALID_VALUE),
```
Comment (VYaniv) [inline src/entities/user/validation/createUserValidation.schema.ts]: Probably required should be the last.

File: src/entities/user/service/user.service.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  async createUser(userData: CreateUserDto): Promise<UserDocument> {
+    const {firstName, lastName, email, phoneNumber, password} = userData;
+    const saltOrRounds = 10;
+    const hash = await bcrypt.hash(password, saltOrRounds);
+    const newUser = new this.userModel({
+      firstName,
+      lastName,
+      email,
+      phoneNumber,
+      passwordHash: hash
```
Comment (VYaniv) [inline src/entities/user/service/user.service.ts]: It potentially can be implemented via setter. When you will hash it in a setter and here just pass a password.

File: src/pipes/YupValidationPipe.util.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import {Schema} from 'yup';
+
+@Injectable()
+export class YupValidationPipe<T> implements PipeTransform<T> {
+  constructor(private readonly schema: Schema<T>) {}
+
+  async transform(value: T) {
+    try {
+      await this.schema.validate(value, {abortEarly: false});
+    } catch (err) {
```
Comment (VYaniv) [inline src/pipes/YupValidationPipe.util.ts]: `err` -> `error`

File: src/entities/user/model/user.schema.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+@Schema()
+export class User {
+  @Prop({required: true})
+  firstName: string;
+
+  @Prop({required: true})
+  lastName: string;
+
+  @Prop({required: true})
```
Comment (VYaniv) [inline src/entities/user/model/user.schema.ts]: email also should be unique.

File: src/entities/user/service/user.service.ts (hunk 4 above)
Comment (denys-rybachok) [reply src/entities/user/service/user.service.ts]: Ok, thanks. Where to place this file with setter better? I assume I don`t want to have it directly in user.schema.ts. So I can create separate file like setPassword.ts in a model folder?

Comment (VYaniv) [reply src/entities/user/service/user.service.ts]: You can skip it for now if you want, it just an idea

Comment (denys-rybachok) [reply src/entities/user/service/user.service.ts]: A problem here. The password is hashed asynchronously. Setters works synchronous only. I will try to solve.

Comment (denys-rybachok) [reply src/entities/user/service/user.service.ts]: Solved.
```
UserSchema.pre('save', async function (next) {
  if (this.isModified('password')) {
    this.password = await setPassword(this.password);
  }
  next();
});
```

Comment (VYaniv) [reply src/entities/user/service/user.service.ts]: Great.

**Review by VYaniv — APPROVED** [review]

**Pull request merged**